### PR TITLE
Allow promise to be returned from 'add()'

### PR DESCRIPTION
**Pull request by dchambers**
FWIW, this is a partial fix for https://github.com/storybooks/storybook/issues/713.

There are no docs, I haven't promisified the other test functions in storyshot (only `snapshotWithOptions`), and there are no unit tests.

### CODE REVIEW

**Comment by kanzelm3** [comment]
+1 this feature would be useful

**Review by ndelangen — APPROVED** [review]
Code looks good

not tested yet

Please resolve linting errors

**Review by danielduan — APPROVED** [review]

**Comment by ndelangen** [comment]
Hey @dchambers code looks good to me, and with a clear use-case described [here](https://github.com/storybooks/storybook/issues/713#issuecomment-323844440) I think we can get this rolling.

I think a unit-test or 2 to test this, a note of this in the docs somewhere would be nice.
An example usage in a kitchen-sink (that is snapshotted) is required to get this merged.

**Review by ndelangen — CHANGES_REQUESTED** [review]
We need an example usage that is snapshotted in a kitchen sink

**Comment by JasonCoal** [comment]
What is this?   
  
  
  

  
  

  
  
>

**Comment by dchambers** [comment]
Hi @ndelangen, I've taken quite a good look at the tests that currently exist and can't find anything that looks like it would be able to verify this new functionality. I guess part of the problem is that what's there at present seems to be more like unit tests rather than BDD tests that test the actual behaviour, and in this case unit testing that some promises were added wouldn't seem to really prove anything.

If I were to add someting then the `app/react/src/client/preview/client_api.test.js` test would seem to be the place for a test of `add()` though I would have thought? Any ideas here?

Regarding docs, I have now added a description of the feature that will hopefully be helpful to people.

Thanks again!

File: docs/pages/basics/writing-stories/index.md [hunk 1]
```diff
@@ -1,7 +1,8 @@
----
+* * *
```
Comment (ndelangen) [inline docs/pages/basics/writing-stories/index.md]: Can you revert these lines?, sorry, it's not your fault, the markdown linter needs a patch, But I haven't found the time to understand what changed/broke.

**Comment by ndelangen** [comment]
I think the fastest way to get coverage and confidence is to add a story to a kitchen sink that actually uses this feature!

**Comment by ndelangen** [comment]
@dchambers Are you interested in finishing this PR by processing the review comments I left for you? Anything you need from me? Anything unclear?

Please let me know, I can take over if you'd like.

**Comment by dchambers** [comment]
Hi @ndelangen, yes apologies, I've been crazy busy with work the last two weeks, including this weekend. I'm fairly confident I'll be able to make some time to work through the issues you've identified on Monday, if not Sunday.

Thanks, Dominic 👍

**Comment by dchambers** [comment]
Apologies again for not progressing this at the start of this week as promised. I've still been crazy busy with work. I've now definitely got time to dedicate to this on Friday and Saturday though, so will get everything resolved then. Thanks for your patience on this one 👍

**Comment by ndelangen** [comment]
Rebased for you 👍 , can you check everything is as it should be?

**Comment by dchambers** [comment]
> Rebased for you 👍 , can you check everything is as it should be?

Thanks @ndelangen. It's not quite working yet, but I've just pushed an attempt at a kitchen-sink that is hopefully in the right direction.

**Comment by Hypnosphi** [comment]
Why a separate kitchen-sink?

**Comment by dchambers** [comment]
> Why a separate kitchen-sink?

Good point @Hypnosphi, some storybook tests where one of the stories uses a promise could just as easily be added to the existing `cra-kitchen-sink`. @ndelangen, do you have an opinion here?

One question though, unlike some of the other kitchen-sinks, the CRA one appears to be testing versioned releases of storybook (e.g. `"@storybook/addon-actions": "^3.2.0",` within `package.json`), rather than the repo itself. Is this intentional, or is there some magic here to make it test the current repo?

**Comment by Hypnosphi** [comment]
This magic is called lerna: https://lernajs.io
It links multiple packages between each other when you run `yarn bootstrap -- --core`

We prefer to keep al examples in one kitchen-sink

**Comment by ndelangen** [comment]
This is broken in combination with decorators... Looking for a possible solution.

**Comment by ndelangen** [comment]
Found a solution, I think

**Review by ndelangen — COMMENTED** [review]
So the decorator stuff needs to be reviewed and tested really carefully!!!

There's a few alternatives:
- Render stories twice to detect if it's an async story
- have an api specific for async stories
- do not support decorators for async stories (harder than you think)

What's the risk here?
Well this changes the order of execution of storyFn and decorators.
the storyFn gets the context object earlier then before this change, the decorators get the context object after. I'm unsure what the impact of this really is.
Context has always been -blegh- to me.

What this would like break/change in behavior is if someone was doing this: (which I haven't seen anyone do)
```js
storiesOf('Aliens')
  .add(context => <Alien name={'jimmy'} stuff={context} />)
```

Please advice @Hypnosphi @storybooks/team

**Comment by ndelangen** [comment]
Obviously unit-tests needs fixing / storyshots needs a fix.

I hope we can have some more eyes looking for on the potential impact on this!
And then I or @dchambers can fix them.

**Comment by tmeasday** [comment]
Is this something we should be considering alongside the mooted changes to the addon API?

**Comment by Hypnosphi** [comment]
What was the initial point of passing `story` to decorators as a function? Maybe it's supposed that the decorator can somehow affect that function's output before calling it?

**Comment by Hypnosphi** [comment]
Maybe in `4.0.0` we should introduce some redux middleware-like API for decorators:
```
addDecorator(context => next => story => ...)
```

Then async stories could be supported with promise decorator:
```
addDecorator(context => next => story => Promise.resolve(story).then(next))
```

**Comment by danielduan** [comment]
Closing this because there hasn't been any activity here for a while. Feel free to reopen when this is ready for discussion again.

**Comment by breathe** [comment]
Having trouble following this -- is there still no support for any form of async story definition?

**Comment by leojh** [comment]
I got confirmation that this feature was never finished. What is the appetite to get this effort started again? Or is there a plan for this to be inherent in the v4 release?

**Comment by husayt** [comment]
I have been trying this in 5.2 and it seems this is still not supported. Seems there is an alternative constructs available for using async data in stories. Maybe someone suggest something?

This is what I tried and it doesn't work as add doesn't support promises
```
storiesOf("Article", module)
  .add("test", async () => {
    let dt= await getDynData("article")
    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```

**Comment by Hypnosphi** [comment]
@husayt in `5.2-beta`, you can try using Storybook Hooks. They a similar to [React Hooks](https://reactjs.org/docs/hooks-intro.html) but work with any framework:

```
import {useState, useEffect} from '@storybook/hooks'

storiesOf("Article", module)
  .add("test", () => {
    const [dt, setDt] = useState('Loading...')
    useEffect(() => {
      let mounted = true
      getDynData("article").then(result => {
        if (mounted) {
          setDt(result)
        }
      })
      return () => {
        mounted = false
      }
    })

    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```

You may want to extract a custom hook to reuse this kind of logic:

```
function useAcyncData(getData, placeholder) {
  const [data, setData] = useState(placeholder)
  useEffect(() => {
    let mounted = true
    getData().then(result => {
      if (mounted) {
        setData(result)
      }
    })
    return () => {
      mounted = false
    }
  })
  return data
}

storiesOf("Article", module)
  .add("test", () => {
    const data = useAsyncData(() => getDynData("article"), 'Loading...')

    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```

**Comment by husayt** [comment]
@Hypnosphi Thank you for direction. I have been playing with this since then, but seems it is not ready yet.

This is what I ended up with:
```
import { useEffect, useState } from "@storybook/client-api"
const dt, setDt
storiesOf("Article", module)
  .add("test", () => {
  console.log("rendering started", dlist)
   ;[dt, setDt] = useState('Loading...')
    useEffect(() => {
      getDynData("article").then(result => {
          setDt(result)
      })
    },[])

    return {
      components: { Article },
      data() {
        return {
          art: {
            link: "test",
            title: dt,
            desc: "test",
          }
        }
      },
      template: `<Article v-bind="art" />`
    }
  })
```
Let me go through issues I faced.

1. Seems this has been updated and I should use `client-api` for import
`import { useEffect, useState } from "@storybook/client-api"`
2. Above example was causing endless loop as `setData` would cause rerender and on each one dt would be undefined. So I had to 
 a. define `dt`, `setDt` outside story, so that value would be retained. 
 b . Also, instead of `mounted` logic which you put to tackle infinite calling, I am passing `[]` as second param to `useEffect` and that ensures its called only once.

But still after all this data doesn't get updated on the Vue component. Vue is not getting updates to `dt`. Seems, there needs to be some kind of bridge to pass reactive data from `dt` to Vue. 

I see  ` console.log("rendering started", dlist)` is being called twice, first time with `..Loading`  and second time with the value from api, but component is only rendered once thus only  mounted once. Updates to `dt` don't make any difference to the component.

So seems like these are Vue related issues. Most probably we are jumping ahead of the line with that, as nowhere it is mentioned that this is expected to work with Vue.

BUt I can see how extremely powerful this can be

**Comment by Hypnosphi** [comment]
1. Sure, sorry
2. Just `b.` should be enough. You still need that `mounted` logic to avoid errors when you switch story before the data arrives

**Pull request closed without merging**